usr/lib/common/mech_md5.c

```diff
@@ -67,81 +128,73 @@ CK_RV md5_hash(STDLL_TokData_t *tokdata,
         return CKR_BUFFER_TOO_SMALL;
     }
 
-    rc = md5_hash_update(tokdata, sess, ctx, in_data, in_data_len);
-    if (rc != CKR_OK) {
-        TRACE_DEVEL("md5_hash_update failed\n");
-        return rc;
-    }
+    if (ctx->context == NULL)
+        return CKR_HOST_MEMORY;
 
-    return md5_hash_final(tokdata, sess, FALSE, ctx, out_data, out_data_len);
+    if (ctx->mech.mechanism == CKM_MD5)
+        return sw_md5_hash(ctx, in_data, in_data_len, out_data,
+                           out_data_len);
+    else
+        return CKR_MECHANISM_INVALID;
 }
 
-
 //
 //
-CK_RV md5_hash_update(STDLL_TokData_t *tokdata,
-                      SESSION *sess,
-                      DIGEST_CONTEXT *ctx,
-                      CK_BYTE *in_data, CK_ULONG in_data_len)
+CK_RV md5_hash_update(STDLL_TokData_t *tokdata, SESSION *sess,
+                      DIGEST_CONTEXT *ctx, CK_BYTE *in_data,
+                      CK_ULONG in_data_len)
 {
-    if (!sess || !ctx || !in_data) {
-        TRACE_ERROR("%s received bad argument(s)\n", __func__);
-        return CKR_FUNCTION_FAILED;
-    }
+    UNUSED(tokdata);
+    UNUSED(sess);
 
-    return ckm_md5_update(tokdata, (MD5_CONTEXT *) ctx->context, in_data,
-                          in_data_len);
-}
+    /* if no data to hash, just return */
```

Comment:
is this the right thing to do ?

Reply:
I think yes:
if we don't return here, sw_MD5_Update() gets called, which in turn calls MD5_Update(). This function is already part of OpenSSL and implemented in md32_common.h as HASH_UPDATE. First statement here is: 

>     if (len == 0)
>         return 1;

So we can shorten this path and return in md5_hash_update(). BTW it's the same construction for SHA1.

Reply:
Im asking because md5 of the empty string is d41d8cd98f00b204e9800998ecf8427e . So this is comouted at final ? and just update does not accept empty strings?

Reply:
Yes, update does not accept/handle empty strings. However, this is true only for md5 and sha1. For example, sha256 in openssl is coded differently and the update routine also processes empty strings. 
I made a quick test with the if-statement removed and, as expected, the opencryptoki digest tests all pass ok. So we may remove this if-stmt, or leave it as it is. 
Background: I adjusted the code in mech_md5.c as much as possible with mech_sha.c for easier maintenance. So this if-stmt came in, because mech_sha.c was coded this way. And it makes sense, because the openssl sw implementations of md5 and sha1 are both given by md32_common.h.
If we decide to remove the if-stmt in mech_md5.c, I would also remove it in mech_sha.c.